### PR TITLE
feat(integrations) Add provider method for external id generation

### PR DESCRIPTION
**Pull request by markstory**
Right now the web ui relies on name == identifier to determine if a repository has been connected or not. This only works for a github and bitbucket, other integrations/plugins have other identifiers that they use in their API calls. Exposing this method allows us to expose the provider's ID system to our webui so it can do safer filtering operations.

Refs #10563

### CODE REVIEW

**Review by macqueen — COMMENTED** [review]
i'm a little worried about calling this `external_id` since the repository model does have an `external_id` field, which is typically not this (i think in github's case, it's the numerical id that is guaranteed not to change)

maybe something like `external_slug` would help differentiate?

**Comment by markstory** [comment]
@macqueen Good point. I'll rename the new-derived property.

File: src/sentry/api/endpoints/organization_integration_repos.py [hunk 1]
```diff
@@ -15,7 +15,16 @@ class OrganizationIntegrationReposEndpoint(OrganizationEndpoint):
     permission_classes = (OrganizationIntegrationsPermission, )
 
     def get(self, request, organization, integration_id):
+        """
```
Comment (lauryndbrown) [inline src/sentry/api/endpoints/organization_integration_repos.py]: Great comment here. :)

**Comment by lauryndbrown** [comment]
@macqueen or @markstory if Github has a repository id then why are we not using it? I thought the issue with github was that it did not have an id we could use for a repository...? Am I misunderstanding the comment here?

File: src/sentry/integrations/bitbucket/repository.py [hunk 2]
```diff
@@ -126,3 +126,6 @@ def compare_commits(self, repo, start_sha, end_sha):
                 installation.raise_error(e)
             else:
                 return self._format_commits(repo, res)
+
+    def repository_external_slug(self, repo):
+        return repo.name
```
Comment (lauryndbrown) [inline src/sentry/integrations/bitbucket/repository.py]: Bitbucket does in fact have a repository id that could be utilized. This could perhaps be done in a different pr if we wanted to change it. We don't have to, however, it'll work as is.

File: src/sentry/integrations/repositories.py [hunk 3]
```diff
@@ -20,6 +20,10 @@ def get_repositories(self, query=None):
             'name': display_name,
             'identifier': external_repo_id,
         }]
+
+        The shape of the `identifier` should match the data
```
Comment (lauryndbrown) [inline src/sentry/integrations/repositories.py]: Thanks for adding this

File: src/sentry/integrations/github/repository.py [hunk 4]
```diff
@@ -142,4 +142,7 @@ def _transform_patchset(self, diff):
         return changes
 
     def pull_request_url(self, repo, pull_request):
-        return u'{}/pulls/{}'.format(repo.url, pull_request.key)
+        return u'{}/pull/{}'.format(repo.url, pull_request.key)
+
+    def repository_external_slug(self, repo):
+        return repo.name
```
Comment (lauryndbrown) [inline src/sentry/integrations/github/repository.py]: For Bitbucket and GitHub, is there a reason this isn't the `external_id`? What happens if the name changes?

**Comment by markstory** [comment]
@lauryndbrown GitHub does have a repository.id however, you can't use it for anything. All of GitHub's API endpoints want the `:owner/:repo` path/slug.

File: tests/sentry/api/endpoints/test_organization_repositories.py [hunk 5]
```diff
@@ -35,6 +35,28 @@ def test_simple(self):
...
+
+        assert response.status_code == 200, response.content
+        assert len(response.data) == 1
+        first_row = response.data[0]
+        assert first_row['id'] == six.text_type(repo.id)
+        assert first_row['provider'] == {
+            'id': 'dummy',
+            'name': 'Example'
+        }
+        assert first_row['externalSlug'] == six.text_type(repo.external_id)
```
Comment (lauryndbrown) [inline tests/sentry/api/endpoints/test_organization_repositories.py]: Since part of the issue was that the id could be a number could you add a test with a numerical/alpha-numerical id?

File: src/sentry/integrations/github/repository.py (hunk 4 above)
Comment (lauryndbrown) [reply src/sentry/integrations/github/repository.py]: I saw your comment above! @markstory

File: src/sentry/integrations/bitbucket/repository.py (hunk 2 above)
Comment (markstory) [reply src/sentry/integrations/bitbucket/repository.py]: We'd also have to update the repository add process for bitbucket to use the numeric id then. These 'slugs' have to match the `identifier` used in the integration repository list API. Its a pretty fragile contract right now, I'd love to revisit and make it harder to accidentally break. This at least moves us away from using fields named `name` for identity comparison.

File: tests/sentry/api/endpoints/test_organization_repositories.py (hunk 5 above)
Comment (markstory) [reply tests/sentry/api/endpoints/test_organization_repositories.py]: This test is checking that the `externalSlug` field matches `repo.external_id` which is a number-y string. Is that what you were thinking?

File: src/sentry/integrations/repositories.py [hunk 6]
```diff
@@ -20,6 +20,10 @@ def get_repositories(self, query=None):
...
+        returned by the integration's
+        IntegrationRepositoryProvider.repository_external_id()
```
Comment (macqueen) [inline src/sentry/integrations/repositories.py]: nit: should be `IntegrationRepositoryProvider.repository_external_slug()` now

File: tests/sentry/integrations/vsts/test_repository.py [hunk 7]
```diff
@@ -117,3 +117,11 @@ def test_build_repository_config(self):
             },
             'integration_id': integration.id,
         }
+
+    def test_repository_external_slug(self):
+        repo = Repository(
+            name='MyFirstProject',
+            url='https://mbittker.visualstudio.com/_git/MyFirstProject/',
```
Comment (macqueen) [inline tests/sentry/integrations/vsts/test_repository.py]: maaaaxxxxx 😢

**Review by macqueen — APPROVED** [review]
✨

File: tests/sentry/api/endpoints/test_organization_repositories.py (hunk 5 above)
Comment (lauryndbrown) [reply tests/sentry/api/endpoints/test_organization_repositories.py]: @markstory I'm sorry; so I think I actually meant to put this on the JS tests! If I understood everything ok in the js tests, you only test: `externalSlug: 'example/repo-name',` Let me know if this isn't the case. And I'm ok with approving.
